source/websocket_encoder.c

```diff
@@ -0,0 +1,347 @@
+        encoder->state = AWS_WEBSOCKET_ENCODER_STATE_MASKING_KEY_CHECK;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+/* MASKING_KEY_CHECK: Outputs no data. Gets things ready for (or decides to skip) the STATE_MASKING_KEY */
+static int s_state_masking_key_check(struct aws_websocket_encoder *encoder, struct aws_byte_buf *out_buf) {
+    (void)out_buf;
+
+    if (encoder->frame.masked) {
+        encoder->state_bytes_processed = 0;
+        encoder->state = AWS_WEBSOCKET_ENCODER_STATE_MASKING_KEY;
+    } else {
+        encoder->state = AWS_WEBSOCKET_ENCODER_STATE_PAYLOAD_CHECK;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+/* MASKING_KEY: Output masking-key (state skipped if no masking key). */
+static int s_state_masking_key(struct aws_websocket_encoder *encoder, struct aws_byte_buf *out_buf) {
+    /* Prepare cursor to iterate over masking-key bytes */
+    struct aws_byte_cursor cursor =
+        aws_byte_cursor_from_array(encoder->frame.masking_key, sizeof(encoder->frame.masking_key));
+
+    /* Advance cursor if some bytes already written */
+    aws_byte_cursor_advance(&cursor, (size_t)encoder->state_bytes_processed);
+
+    /* Shorten cursor if it won't all fit in out_buf */
+    bool all_data_written = true;
+    size_t space_available = out_buf->capacity - out_buf->len;
+    if (cursor.len > space_available) {
+        cursor.len = space_available;
+        all_data_written = false;
+    }
+
+    aws_byte_buf_write_from_whole_cursor(out_buf, cursor);
+    encoder->state_bytes_processed += cursor.len;
+
+    /* If all bytes written, advance to next state */
+    if (all_data_written) {
+        encoder->state = AWS_WEBSOCKET_ENCODER_STATE_PAYLOAD_CHECK;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+/* MASKING_KEY_CHECK: Outputs no data. Gets things ready for (or decides to skip) STATE_PAYLOAD */
+static int s_state_payload_check(struct aws_websocket_encoder *encoder, struct aws_byte_buf *out_buf) {
+    (void)out_buf;
+
+    if (encoder->frame.payload_length > 0) {
+        encoder->state_bytes_processed = 0;
+        encoder->state = AWS_WEBSOCKET_ENCODER_STATE_PAYLOAD;
+    } else {
+        encoder->state = AWS_WEBSOCKET_ENCODER_STATE_DONE;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+/* PAYLOAD: Output payload until we're done (state skipped if no payload). */
+static int s_state_payload(struct aws_websocket_encoder *encoder, struct aws_byte_buf *out_buf) {
+
+    /* Bail early if out_buf has no space for writing */
+    if (out_buf->len >= out_buf->capacity) {
+        return AWS_OP_SUCCESS;
+    }
+
+    const uint64_t prev_bytes_processed = encoder->state_bytes_processed;
+    const struct aws_byte_buf prev_buf = *out_buf;
+
+    /* Invoke callback which will write to buffer */
+    bool user_says_done = false;
+    int err = encoder->stream_outgoing_payload(out_buf, &user_says_done, encoder->user_data);
+    if (err) {
+        return AWS_OP_ERR;
+    }
+
+    /* Ensure that user did not commit forbidden acts upon the out_buf */
```

Comment:
:+1:

Is a crash the only way out here? Could we log something?

Reply:
Yeah, this should get an FATAL log. I'm going to add logging in a subsequent pass